### PR TITLE
Make autograd profiler thread-local

### PR DESCRIPTION
**Pull request by jamesr66a**
This should be no-op for Python.

This allow us to, for example, record request-level traces in a production service. Threads in the autograd engine as well as the c10 thread pool inherit their "profiler state" from the thread that issued the work.

### CODE REVIEW

File: torch/csrc/autograd/profiler.cpp [hunk 1]
```diff
@@ -16,35 +16,55 @@ TORCH_API void registerCUDAMethods(CUDAStubs* stubs) {
...
-    event_list = std::make_shared<RangeEventList>();
-    thread_id = next_thread_id++;
-    all_event_lists.emplace_front(event_list);
+  if (!event_list || associated_invocation_state != invocation_state.get()) {
+    std::lock_guard<std::mutex> guard(invocation_state->mutex);
+    if (thread_id == -1) {
+      thread_id = next_thread_id++;
+    }
+    if (!invocation_state->all_event_lists[thread_id]) {
+      event_list = invocation_state->all_event_lists[thread_id] = std::make_shared<RangeEventList>();
```
Comment (dzhulgakov) [inline torch/csrc/autograd/profiler.cpp]: if event_list is a local cache - don't you need to update it here even if no new list needs to be created?

**Review by dzhulgakov — COMMENTED** [review]
shall we add a test for this logic?

**Review by zdevito — COMMENTED** [review]
We are going to need to figure out a way to support both profiling across all threads and recording information about a single thread. Would it be possible to simply record _everything_ and then have each event aware of the originator thread id in addition to the current id of the running thread? That is, we separate the physical thread id where a task runs from the thread id where the interpreter first ran that caused the forked process to be created.

File: torch/csrc/autograd/profiler.cpp (hunk 1 above)
Comment (jamesr66a) [reply torch/csrc/autograd/profiler.cpp]: `event_list` is a pointer into the list owned by the `ProfilerInvocationState`. Only the thread-local state needs to be updated to point to the right thing

File: torch/csrc/autograd/profiler.cpp [hunk 2]
```diff
@@ -16,35 +16,55 @@ TORCH_API void registerCUDAMethods(CUDAStubs* stubs) {
...
+// Assign all threads contiguous integer IDs so we can keep track of them
+std::atomic<int> next_thread_id{0};
+thread_local int thread_id = -1;
+
+thread_local std::shared_ptr<ProfilerInvocationState> invocation_state =
+    std::make_shared<ProfilerInvocationState>();
+
 thread_local std::shared_ptr<RangeEventList> event_list;
-thread_local uint16_t thread_id;
+thread_local ProfilerInvocationState* associated_invocation_state = invocation_state.get();
```
[the remark this answers is not in the document]
Comment (jamesr66a) [reply torch/csrc/autograd/profiler.cpp]: This is basically an indicator for if `invocation_state` was changed. If so we need to update our thread-local event list to point to one owned by the current `invocation_state`. I did it this way so we don't have to acquire the lock on `invocation_state` every time we call `getEventList`

File: torch/csrc/autograd/profiler.cpp [hunk 3]
```diff
@@ -16,35 +16,55 @@ TORCH_API void registerCUDAMethods(CUDAStubs* stubs) {
...
-uint16_t next_thread_id = 0;
-std::mutex all_event_lists_mutex;
-std::list<std::shared_ptr<RangeEventList>> all_event_lists;
+// Assign all threads contiguous integer IDs so we can keep track of them
+std::atomic<int> next_thread_id{0};
+thread_local int thread_id = -1;
+
+thread_local std::shared_ptr<ProfilerInvocationState> invocation_state =
+    std::make_shared<ProfilerInvocationState>();
+
```
[the remark this answers is not in the document]
Comment (jamesr66a) [reply torch/csrc/autograd/profiler.cpp]: Let's talk offline, I don't think we have the same mental model for what's going on here

[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/autograd/profiler.cpp]: Can you please also post a summary online? 😊

File: torch/csrc/autograd/profiler.cpp [hunk 4]
```diff
@@ -16,35 +16,55 @@ TORCH_API void registerCUDAMethods(CUDAStubs* stubs) {
...
 thread_local std::shared_ptr<RangeEventList> event_list;
-thread_local uint16_t thread_id;
+thread_local ProfilerInvocationState* associated_invocation_state = invocation_state.get();
+
+WorkerPushProfileState::WorkerPushProfileState(std::shared_ptr<ProfilerInvocationState> st) {
+  old_state = invocation_state;
+  invocation_state = st;
```
Comment (apaszke) [inline torch/csrc/autograd/profiler.cpp]: nit: move

File: torch/csrc/autograd/profiler.cpp [hunk 5]
```diff
@@ -16,35 +16,55 @@ TORCH_API void registerCUDAMethods(CUDAStubs* stubs) {
...
+}
+
+WorkerPushProfileState::~WorkerPushProfileState() {
+  invocation_state = old_state;
```
Comment (apaszke) [inline torch/csrc/autograd/profiler.cpp]: Eh I guess we should start distinguishing globals somehow

**Review by apaszke — COMMENTED** [review]
Can you please make sure this doesn't regress the perf of our profiler before you land this? We had some regressions in the past, and it's quite annoying when it gets slow.

also a nit: can we just drop the `invocation` part? Profiler state seems like a reasonable description.

**Pull request closed without merging**